Open Judge System/Services/OJS.Services/Business/ParticipantScores/ParticipantScoresBusinessService.cs

```diff
@@ -44,6 +59,146 @@ public void NormalizeAllPointsThatExceedAllowedLimit()
             }
         }
 
+        public CategoryContestsParticipationSummary GetCategoryParticipationSummary(int categoryId, bool showHidden, bool official = true)
+        {
+            var contests = this.contestsDataService
+                .GetAllNotDeletedByCategory(categoryId, showHidden)
+                .Include(c => c.ProblemGroups)
+                .Include(c => c.ProblemGroups.Select(pg => pg.Problems))
+                .OrderByDescending(c => c.CreatedOn)
+                .ToList();
+
+            var categoryResults = contests
+                .Select(c => this.GetParticipationSummaryForContest(c, official))
+                .ToList();
+            
+            var maxProblemsCount = categoryResults
+                .Select(ps => ps.ProblemsCount)
+                .Max();
+
+            return new CategoryContestsParticipationSummary
+            {
+                MaxProblemsCount = maxProblemsCount,
+                Results = categoryResults,
+            };
+        }
+        
+        public ParticipationsSummaryServiceModel GetParticipationSummaryForContest(Contest contest, bool official)
+        {
+            var participants = 
+                        participantsData.GetAllByContestAndIsOfficial(contest.Id, official)
+                            .Include(p => p.User)
+                            .Include(p => p.Submissions)
+                            .Include(p => p.Scores)
+                            .Include(p => p.Problems)
+                    .ToList();
+            
+            var results = participants
+                .Where(p => p.Scores.Count != 0 && p.Scores.All(ps => ps.Submission != null))
+                .Select(this.SummarizeParticipationWithMaximumPoints)
+                .ToList();
+
+            var problemsCount = contest.ProblemGroups.Count(pg => !pg.Problems.IsNullOrEmpty());
+            
+            return new ParticipationsSummaryServiceModel
+            {
+                ContestId = contest.Id,
+                ContestName = contest.Name,
+                Results = results,
+                ProblemsCount = problemsCount
+            };
+        }
+
+        private ParticipantScoresSummaryModel SummarizeParticipationWithMaximumPoints(Participant participant)
+        {
+            Dictionary<int, DateTime> problemOrderToFirstBestSubmissionTimeCreated = this.CalculateFirstBestSubmissionTimeForProblems(participant);
+
+            var userStartTime = participant.ParticipationStartTime ?? participant.CreatedOn;
+
+            SortedDictionary<int, double> problemOrderToTimeTakenBetweenBest =
```

Comment:
Is this explicit `SortedDictionary<int, double>` necessary here?

Reply:
do not see harm, I read it better this way

Reply:
do not see harm, I read it better this way